PR: https://github.com/odoo/odoo/pull/215599

From: e992fcff101f7e7f05efd0d755f8a28541c525e0
From: rmun-odoo
Date: 2026-02-25 11:04:17

Breaking data model changes scores: del:2 + add:0, change matches:
-    owner_user_id = fields.Many2one(compute='_compute_owner', store=True)
-    assign_date = fields.Date(compute='_compute_equipment_assign', store=True, readonly=False, copy=True)

Total Changes: 72

[IMP] maintenace, hr_maintenance: move equipment_assign_to field

* Moved equipment_assign_to field from hr_maintenance to maintenance
  so that we can add different selection option to assign equipment from
  hr_maintenance as well as fleet_maintenance modules.
* Updated the existing `owner_user_id` field to be stored and computed in the
  maintenance module. This enables extending custom owner computation logic in
 `hr_maintenance` and `fleet_maintenance`.
* Added a custom widget for the `equipment_assigned_to` field so that all
  options are displayed as desired. Adding in `maintenance` so that we do not
  have to add it twice in both `hr_maintenance` and `fleet_maintenance`.

task-485543

Part-of: odoo/odoo#215599
Related: odoo/upgrade#9562
Signed-off-by: Yannick Tivisse (yti) <yti@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr_maintenance/models/equipment.py
+++ b/addons/hr_maintenance/models/equipment.py
@@ -1,45 +1,25 @@
 # -*- coding: utf-8 -*-
 
 from odoo import api, fields, models, tools
+from odoo.fields import Domain
 
 
 class MaintenanceEquipment(models.Model):
     _inherit = 'maintenance.equipment'
 
-    employee_id = fields.Many2one('hr.employee', compute='_compute_equipment_assign',
+    employee_id = fields.Many2one('hr.employee', compute='_compute_equipment_assignment_fields',
         store=True, readonly=False, string='Assigned Employee', tracking=True, index='btree_not_null')
-    department_id = fields.Many2one('hr.department', compute='_compute_equipment_assign',
+    department_id = fields.Many2one('hr.department', compute='_compute_equipment_assignment_fields',
         store=True, readonly=False, string='Assigned Department', tracking=True)
     equipment_assign_to = fields.Selection(
-        [('department', 'Department'), ('employee', 'Employee'), ('other', 'Other')],
-        string='Used By',
-        required=True,
-        default='employee')
-    owner_user_id = fields.Many2one(compute='_compute_owner', store=True)
-    assign_date = fields.Date(compute='_compute_equipment_assign', store=True, readonly=False, copy=True)
+        selection_add=[('department', 'Department'), ('employee', 'Employee')], required=True,
+        ondelete={'department': 'set other', 'employee': 'set other'}, default='employee')
 
-    @api.depends('employee_id', 'department_id', 'equipment_assign_to')
-    def _compute_owner(self):
-        for equipment in self:
-            equipment.owner_user_id = self.env.user.id
-            if equipment.equipment_assign_to == 'employee':
-                equipment.owner_user_id = equipment.employee_id.user_id.id
-            elif equipment.equipment_assign_to == 'department':
-                equipment.owner_user_id = equipment.department_id.manager_id.user_id.id
-
-    @api.depends('equipment_assign_to')
-    def _compute_equipment_assign(self):
+    @api.depends(lambda self: self._get_assign_fields())
+    def _compute_is_assigned(self):
+        assign_fields = self._get_assign_fields()
         for equipment in self:
-            if equipment.equipment_assign_to == 'employee':
-                equipment.department_id = False
-                equipment.employee_id = equipment.employee_id
-            elif equipment.equipment_assign_to == 'department':
-                equipment.employee_id = False
-                equipment.department_id = equipment.department_id
-            else:
-                equipment.department_id = equipment.department_id
-                equipment.employee_id = equipment.employee_id
-            equipment.assign_date = fields.Date.context_today(self)
+            equipment.is_assigned = any(equipment[field] for field in assign_fields)
 
     @api.model_create_multi
     def create(self, vals_list):
@@ -77,6 +57,40 @@ class MaintenanceEquipment(models.Model):
             return self.env.ref('maintenance.mt_mat_assign')
         return super(MaintenanceEquipment, self)._track_subtype(init_values)
 
+    def _get_assign_fields(self):
+        return super()._get_assign_fields() + ['employee_id', 'department_id']
+
+    def _get_owner_methods_by_equipment_assign_to(self):
+        owner_methods = super()._get_owner_methods_by_equipment_assign_to()
+        owner_methods.update({
+            'employee': lambda eq: eq.employee_id.user_id.id or self.env.user.id,
+            'department': lambda eq: eq.department_id.manager_id.user_id.id or self.env.user.id,
+        })
+        return owner_methods
+
+    def _get_assignment_handlers_by_equipment_assign_to(self):
+        handlers = super()._get_assignment_handlers_by_equipment_assign_to()
+        handlers.update({
+            'employee': lambda eq: {
+                field: eq[field] if field == 'employee_id' else False
+                for field in self._get_assign_fields()
+            },
+            'department': lambda eq: {
+                field: eq[field] if field == 'department_id' else False
+                for field in self._get_assign_fields()
+            },
+        })
+        return handlers
+
+    def _search_is_assigned(self, operator, value):
+        if operator not in ('=', '!=') or value not in (True, False):
+            return NotImplemented
+        assign_fields = self._get_assign_fields()
+        is_equipment_assigned = (operator == "=") == value
+        if is_equipment_assigned:
+            return Domain.OR(Domain(field, "!=", False) for field in assign_fields)
+        return Domain.AND(Domain(field, "=", False) for field in assign_fields)
+
 
 class MaintenanceRequest(models.Model):
     _inherit = 'maintenance.request'
